aptos-move/framework/aptos-framework/sources/multisig_account.move

```diff
@@ -0,0 +1,1476 @@
+        simple_map::add(&mut transaction.votes, creator, true);
+
+        let sequence_number = multisig_account.next_sequence_number;
+        multisig_account.next_sequence_number = sequence_number + 1;
+        table::add(&mut multisig_account.transactions, sequence_number, transaction);
+        emit_event(
+            &mut multisig_account.create_transaction_events,
+            CreateTransactionEvent { creator, sequence_number, transaction },
+        );
+    }
+
+    fun create_multisig_account(owner: &signer): (signer, SignerCapability) {
+        let owner_nonce = account::get_sequence_number(address_of(owner));
+        let (multisig_signer, multisig_signer_cap) =
+            account::create_resource_account(owner, create_multisig_account_seed(to_bytes(&owner_nonce)));
+        // Register the account to receive APT as this is not done by default as part of the resource account creation
+        // flow.
+        if (!coin::is_account_registered<AptosCoin>(address_of(&multisig_signer))) {
+            coin::register<AptosCoin>(&multisig_signer);
+        };
+
+        (multisig_signer, multisig_signer_cap)
+    }
+
+    fun create_multisig_account_seed(seed: vector<u8>): vector<u8> {
+        // Generate a seed that will be used to create the resource account that hosts the staking contract.
+        let multisig_account_seed = vector::empty<u8>();
+        vector::append(&mut multisig_account_seed, DOMAIN_SEPARATOR);
+        vector::append(&mut multisig_account_seed, seed);
+
+        multisig_account_seed
+    }
+
+    fun validate_owners(owners: &vector<address>, multisig_account: address) {
+        let distinct_owners = simple_map::create<address, bool>();
+        let i = 0;
+        let len = vector::length(owners);
+        while (i < len) {
+            let owner = *vector::borrow(owners, i);
+            assert!(owner != multisig_account, error::invalid_argument(EOWNER_CANNOT_BE_MULTISIG_ACCOUNT_ITSELF));
+            assert!(
+                !simple_map::contains_key(&distinct_owners, &owner),
+                error::invalid_argument(EDUPLICATE_OWNER),
+            );
+            simple_map::add(&mut distinct_owners, owner, true);
+            i = i + 1;
+        }
+    }
+
+    fun assert_is_owner(owner: &signer, multisig_account: &MultisigAccount) {
+        assert!(
+            vector::contains(&multisig_account.owners, &address_of(owner)),
+            error::permission_denied(ENOT_OWNER),
+        );
+    }
+
+    fun num_approvals_and_rejections(owners: &vector<address>, transaction: &MultisigTransaction): (u64, u64) {
+        let num_approvals = 0;
+        let num_rejections = 0;
+
+        let votes = &transaction.votes;
+        let len = vector::length(owners);
+        let i = 0;
+        while (i < len) {
+            let owner = *vector::borrow(owners, i);
+            i = i + 1;
+
+            if (!simple_map::contains_key(votes, &owner)) {
```

Comment:
this n^2 lookup is ugly, probably can get better with inline functions later

Reply:
Yeah unfortunately. We can also do better with storing owners with simple_set whenever https://github.com/aptos-labs/aptos-core/pull/5920 is merged.